frontend/src/pages/Compare.tsx

```diff
@@ -14,56 +14,110 @@
  * limitations under the License.
  */
 
-import React from 'react';
+import React, { useEffect, useState } from 'react';
 import { useQuery } from 'react-query';
 import { ApiRunDetail } from 'src/apis/run';
 import { QUERY_PARAMS } from 'src/components/Router';
 import { FeatureKey, isFeatureEnabled } from 'src/features';
 import { Apis } from 'src/lib/Apis';
 import { errorToMessage } from 'src/lib/Utils';
 import { URLParser } from '../lib/URLParser';
-import CompareV1, { CompareState } from './CompareV1';
+import CompareV1 from './CompareV1';
 import CompareV2 from './CompareV2';
 import { PageProps } from './Page';
 
-export type CompareProps = PageProps & CompareState;
+enum CompareVersion {
+  V1,
+  V2,
+  Mixed,
+  InvalidRunCount,
+  Unknown,
+}
 
 // This is a router to determine whether to show V1 or V2 compare page.
-export default function Compare(props: CompareProps) {
+export default function Compare(props: PageProps) {
+  const { updateBanner } = props;
+  const [compareVersion, setCompareVersion] = useState<CompareVersion>(CompareVersion.Unknown);
   const queryParamRunIds = new URLParser(props).get(QUERY_PARAMS.runlist);
   const runIds = (queryParamRunIds && queryParamRunIds.split(',')) || [];
 
-  // Retrieves run details.
-  const { isSuccess, isError, data } = useQuery<ApiRunDetail[], Error>(
+  // Retrieves run details, set page version on success.
+  const { isError, error, data } = useQuery<ApiRunDetail[], Error>(
     ['run_details', { ids: runIds }],
     () => Promise.all(runIds.map(async id => await Apis.runServiceApi.getRun(id))),
     {
       staleTime: Infinity,
-      onError: async error => {
+    },
+  );
+
+  useEffect(() => {
+    // Set the version based on the runs included.
+    if (data) {
+      if (data.length < 2 || data.length > 10) {
+        setCompareVersion(CompareVersion.InvalidRunCount);
+      } else {
+        const v2runs = data.filter(run =>
+          run.run?.pipeline_spec?.hasOwnProperty('pipeline_manifest'),
+        );
+        if (v2runs.length === 0) {
+          setCompareVersion(CompareVersion.V1);
+        } else if (v2runs.length === data.length) {
+          setCompareVersion(CompareVersion.V2);
+        } else {
+          setCompareVersion(CompareVersion.Mixed);
+        }
+      }
+    }
+  }, [data]);
+
+  useEffect(() => {
+    // Update banner based on error, feature flag, run versions, and run count.
+    if (isError) {
+      (async function() {
```

Comment:
I fear that we are unnecessarily introducing async function within useEffect, which is dangerous because it doesn't guarantee that the latest information is applied to the banner. Can we just use `const errorMessage = error.message` instead?

Reply:
I can see that's an important concern, as it could wipe out the correct banner information. I tested this with a purposefully non-existent run (an ID of `1`, as shown in the URL), and it doesn't show the error details.

| Original Page (using `await errorToMessage(error)`) | Non-Async Page (using `error.message`) |
| ------------- | ------------- |
| ![ErrorWithoutDetails](https://user-images.githubusercontent.com/7987279/172919978-fea0f89a-b52e-4731-873b-70cbb9d77450.png) | ![ErrorWithDetails](https://user-images.githubusercontent.com/7987279/172920810-065e79ed-7dbc-4cb7-bee6-d8fe4dce1a56.png) |

I'm still happy to show this if you think this is the better approach, as the original could run into async issues.

This research actually brought up a new issue though - since `isError` is `false` when the request is being sent but unsuccessful during initial attempts, the error banner disappears when the `useQuery` attempts to re-fetch the data (when occurs when you bring the page into focus again, as shown in the below screencasts).

**Original Page with Error**

https://user-images.githubusercontent.com/7987279/172939214-e274a5c9-fad6-46aa-950a-56d7938c0fc3.mp4

I fixed this by adding a check that prevented any banner from loading while `isLoading === true`; therefore, when the data is being fetched / loaded again, the banner is not instantly cleared until some value is actually returned.

**New Page with `isLoading` Check**

https://user-images.githubusercontent.com/7987279/172939231-d2e5a34b-ff38-4802-b684-cecf17a8f78b.mp4

I made this fix at this commit: https://github.com/kubeflow/pipelines/pull/7844/commits/ff50bb8144a0eec347935e80befcebfdb25923cf.

Reply:
That is awesome research! I agree with your approach here, just a small NIT: 

Since you are checking `isLoading`, you need to add `isLoading` to your `useEffect`:

```
useEffect(() => {
 ...
}, [compareVersion, isLoading, isError, error, updateBanner, runIds.length]);
```

But because the isLoading and isError usually change at the same time, it doesn't make much difference in practice regardless of having `isLoading` key in the `useEffect`. We can do it in the future PR.

Reply:
Ah, sorry I missed that - sounds good, I'll add it in the next PR. Thanks!

Reply:
I added this dependency here: https://github.com/kubeflow/pipelines/pull/7882/files#r895048806.